src/history/hs.c

```diff
@@ -394,22 +394,54 @@ __hs_insert_updates_verbose(WT_SESSION_IMPL *session, WT_BTREE *btree)
  */
 static int
 __hs_insert_record_with_btree_int(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BTREE *btree,
-  const WT_ITEM *key, const WT_UPDATE *upd, const uint8_t type, const WT_ITEM *hs_value,
+  WT_ITEM *key, const WT_UPDATE *upd, const uint8_t type, const WT_ITEM *hs_value,
   WT_HS_TIME_POINT *stop_time_point)
 {
     WT_CURSOR_BTREE *cbt;
+    WT_DECL_ITEM(hs_key);
     WT_DECL_RET;
     WT_UPDATE *hs_upd, *upd_local;
+    wt_timestamp_t hs_start_ts;
+    uint64_t counter, hs_counter;
+    uint32_t hs_btree_id;
+    int cmp;
 
     cbt = (WT_CURSOR_BTREE *)cursor;
     hs_upd = upd_local = NULL;
 
+    /* Allocate buffers for the data store and history store key. */
+    WT_ERR(__wt_scr_alloc(session, 0, &hs_key));
+
+    /*
+     * Adjust counter if there exists an update in the history store with same btree id, key and
+     * timestamp. Otherwise the newly inserting history store record may fall behind the existing
+     * one can lead to wrong order.
+     */
+    counter = __wt_atomic_add64(&btree->hs_counter, 1);
```

Comment:
Does 'btree->hs_counter` provide any value with this solution?  It looks like you could set `counter = 0` here and get the same effect...

Reply:
I thought of comparing the counter with retrieved hs_counter value along with btree id and timestamp before going into comparing the key.

`if (hs_btree_id == btree->id && upd->start_ts == hs_start_ts && counter < hs_counter)`

As these comparisons are less costly than the key. If the counter value is already more than the existing key, there is no need to compare the key even if the same key with the same timestamp data already exists in the history store.

Reply:
I don't think the call to `__wt_compare` there will be noticable -- we've don't lots of comparisons already during `__wt_hs_cursor_position`.  What I'm concerned about is that it takes 1 byte to store a zero counter value, but grow the key by an extra 4-8 bytes to store `btree->hs_counter` depending on how long the system stays up.

Reply:
Fixed by removing `btree->hs_counter`. The counter value in the key can be 0 for different timestamp updates, in case if there is the same timestamp update that is already present, the value can max counter + 1.